pkg/instance/handler_test.go

```diff
@@ -0,0 +1,142 @@
+package instance
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"net/http"
+	"net/http/httptest"
+	"testing"
+
+	"github.com/dhis2-sre/im-manager/pkg/config"
+
+	"github.com/dhis2-sre/im-manager/pkg/model"
+	"github.com/dhis2-sre/im-user/swagger/sdk/models"
+	"github.com/gin-gonic/gin"
+	"github.com/stretchr/testify/mock"
+	"github.com/stretchr/testify/require"
+	"gorm.io/gorm"
+)
+
+func TestHandler_ListInstances(t *testing.T) {
+	repository := &mockRepository{}
+	groups := []*models.Group{
+		{Name: "group name"},
+	}
+	groupsWithInstances := []GroupWithInstances{
+		{
+			Name: "group name",
+			Instances: []*model.Instance{
+				{
+					Model:     gorm.Model{ID: 1},
+					Name:      "instance name",
+					GroupName: "group name",
+				},
+			},
+		},
+	}
+	repository.
+		On("FindByGroups", groups, false).
+		Return(groupsWithInstances, nil)
+	service := NewService(config.Config{}, repository, nil, nil, nil)
+	handler := NewHandler(nil, service, nil)
+
+	w := httptest.NewRecorder()
+	c := newContext(w, "group name")
+
+	handler.ListInstances(c)
+
+	require.Empty(t, c.Errors)
+	assertResponse(t, w, http.StatusOK, groupsWithInstances)
+	repository.AssertExpectations(t)
+}
+
+func TestHandler_ListInstances_RepositoryError(t *testing.T) {
+	groups := []*models.Group{
+		{Name: "group name"},
+	}
+	repository := &mockRepository{}
+	repository.
+		On("FindByGroups", groups, false).
+		Return(nil, errors.New("some error"))
+	service := NewService(config.Config{}, repository, nil, nil, nil)
+	handler := NewHandler(nil, service, nil)
+
+	w := httptest.NewRecorder()
+	c := newContext(w, "group name")
+
+	handler.ListInstances(c)
+
+	require.Empty(t, w.Body.Bytes())
+	require.Len(t, c.Errors, 1)
+	require.ErrorContains(t, c.Errors[0].Err, "some error")
+	repository.AssertExpectations(t)
+}
+
+func newContext(w *httptest.ResponseRecorder, group string) *gin.Context {
+	user := &models.User{
+		ID: uint64(1),
```

Comment:
I would pass this ID in to the function since this ID needs to match the one in the test setup. Right now that is not obvious 🙂

Reply:
No :smile: That's the user id, it doesn't need to match anything in these tests